src/cost_estimates/fee_medians.rs

```diff
@@ -0,0 +1,359 @@
+
+        Ok(())
+    }
+
+    fn get_rate_estimates(&self) -> Result<FeeRateEstimate, EstimatorError> {
+        Self::get_rate_estimates_from_sql(&self.db, self.window_size)
+    }
+}
+
+/// Computes a `FeeRateEstimate` based on `sorted_fee_rates` using a "weighted percentile" method
+/// described in https://en.wikipedia.org/wiki/Percentile#Weighted_percentile
+///
+/// The percentiles computed are [0.05, 0.5, 0.95].
+///
+/// `sorted_fee_rates` must be non-empty.
+pub fn fee_rate_estimate_from_sorted_weighted_fees(
+    sorted_fee_rates: &Vec<FeeRateAndWeight>,
+) -> FeeRateEstimate {
+    if sorted_fee_rates.is_empty() {
+        panic!("`sorted_fee_rates` cannot be empty.");
+    }
+
+    let mut total_weight = 0f64;
+    for rate_and_weight in sorted_fee_rates {
+        total_weight += rate_and_weight.weight as f64;
+    }
+
+    if total_weight <= 0f64 {
+        panic!("`total_weight` is 0f64. This is an error.");
+    }
+
+    let mut cumulative_weight = 0f64;
+    let mut percentiles = Vec::new();
+    for rate_and_weight in sorted_fee_rates {
+        cumulative_weight += rate_and_weight.weight as f64;
+        let percentile_n: f64 =
+            (cumulative_weight as f64 - rate_and_weight.weight as f64 / 2f64) / total_weight as f64;
+        percentiles.push(percentile_n);
+    }
+    assert_eq!(percentiles.len(), sorted_fee_rates.len());
+
+    let target_percentiles = vec![0.05, 0.5, 0.95];
+    let mut fees_index = 0; // index into `sorted_fee_rates`
+    let mut values_at_target_percentiles = Vec::new();
+    for target_percentile in target_percentiles {
+        while fees_index < percentiles.len() && percentiles[fees_index] < target_percentile {
+            fees_index += 1;
+        }
+        let v = if fees_index == 0 {
+            sorted_fee_rates[0].fee_rate
+        } else if fees_index == percentiles.len() {
+            sorted_fee_rates.last().unwrap().fee_rate
+        } else {
+            // Notation mimics https://en.wikipedia.org/wiki/Percentile#Weighted_percentile
+            let vk = sorted_fee_rates[fees_index - 1].fee_rate;
+            let vk1 = sorted_fee_rates[fees_index].fee_rate;
+            let pk = percentiles[fees_index - 1];
+            let pk1 = percentiles[fees_index];
+            vk + (target_percentile - pk) / (pk1 - pk) * (vk1 - vk)
+        };
+        values_at_target_percentiles.push(v);
+    }
+
+    FeeRateEstimate {
+        high: values_at_target_percentiles[2],
+        middle: values_at_target_percentiles[1],
+        low: values_at_target_percentiles[0],
+    }
+}
+
+/// If the weights in `working_rates` do not add up to `full_block_weight`, add a new entry **in
+/// place** that takes up the remaining space.
+fn maybe_add_minimum_fee_rate(working_rates: &mut Vec<FeeRateAndWeight>, full_block_weight: u64) {
+    let mut total_weight = 0u64;
+    for rate_and_weight in working_rates.into_iter() {
+        total_weight = match total_weight.checked_add(rate_and_weight.weight) {
```

Comment:
Could we realistically encounter an overflow here?  If so, would it make more sense to use `.saturating_add()`?

Reply:
Per the blockchain meeting, the code with `.checked_add()` is fine.  I didn't know that `.saturating_add()` wasn't yet stable for `f64`.  In general, if you're dealing with data whose values can be influenced by the network, you will want to use checked math operations (otherwise, you're inviting a DoS vulnerability).  "Basically impossible" isn't nearly as strong of a guarantee as "impossible" to someone who's motivated to make an unlikely but catastrophic thing happen.